python/ray/data/datasource/image_folder_datasource.py

```diff
@@ -106,22 +119,29 @@ def create_reader(
             filesystem=filesystem,
             root=root,
             size=size,
+            mode=mode,
         )
 
     def _read_file(
-        self, f: "pyarrow.NativeFile", path: str, root: str, size: Tuple[int, int]
+        self,
+        f: "pyarrow.NativeFile",
+        path: str,
+        root: str,
+        size: Optional[Tuple[int, int]],
+        mode: Optional[str],
     ):
-        import imageio as iio
-        import pandas as pd
-        import skimage
+        from PIL import Image
 
         records = super()._read_file(f, path, include_paths=True)
         assert len(records) == 1
         path, data = records[0]
 
-        image = iio.imread(data)
-        image = skimage.transform.resize(image, size)
-        image = skimage.util.img_as_ubyte(image)
+        image = Image.open(io.BytesIO(data))
+        if size is not None:
+            height, width = size
+            image = image.resize((width, height))
+        if mode is not None:
+            image = image.convert(mode)
```

Comment:
how about adding a unit test for illegal `mode`, and see if the error message is user-friendly? If not, we should catch here and give a better error message.

Reply:
I think the existing error message is clear enough, but let me know what you think.

Here's the error:

```
>>> ds = ray.data.read_datasource(ImageFolderDatasource(), root="python/ray/data/tests/image-folders/different-modes", mode="X")
2022-08-01 10:40:41,908 WARNING read_api.py:291 -- ⚠️  The number of blocks in this dataset (3) limits its parallelism to 3 concurrent tasks. This is much less than the number of available CPU slots in the cluster. Use `.repartition(n)` to increase the number of dataset blocks.
(_execute_read_task pid=2023) 2022-08-01 10:40:41,913   INFO worker.py:754 -- Task failed with retryable exception: TaskID(3d3e27c54ed1f5cfffffffffffffffffffffffff01000000).
(_execute_read_task pid=2023) Traceback (most recent call last):
(_execute_read_task pid=2023)   File "/Users/bveeramani/GitHub/ray/.venv/lib/python3.9/site-packages/PIL/Image.py", line 1030, in convert
(_execute_read_task pid=2023)     im = self.im.convert(mode, dither)
(_execute_read_task pid=2023) ValueError: conversion from 1 to X not supported
(_execute_read_task pid=2023) 
(_execute_read_task pid=2023) During handling of the above exception, another exception occurred:
(_execute_read_task pid=2023) 
(_execute_read_task pid=2023) Traceback (most recent call last):
(_execute_read_task pid=2023)   File "python/ray/_raylet.pyx", line 662, in ray._raylet.execute_task
(_execute_read_task pid=2023)     with ray._private.worker._changeproctitle(title, next_title):
(_execute_read_task pid=2023)   File "python/ray/_raylet.pyx", line 666, in ray._raylet.execute_task
(_execute_read_task pid=2023)     outputs = function_executor(*args, **kwargs)
(_execute_read_task pid=2023)   File "/Users/bveeramani/GitHub/ray/python/ray/data/_internal/lazy_block_list.py", line 541, in _execute_read_task
(_execute_read_task pid=2023)     block = task()
(_execute_read_task pid=2023)   File "/Users/bveeramani/GitHub/ray/python/ray/data/datasource/datasource.py", line 202, in __call__
(_execute_read_task pid=2023)     for block in result:
(_execute_read_task pid=2023)   File "/Users/bveeramani/GitHub/ray/python/ray/data/datasource/file_based_datasource.py", line 429, in read_files
(_execute_read_task pid=2023)     for data in read_stream(f, read_path, **reader_args):
(_execute_read_task pid=2023)   File "/Users/bveeramani/GitHub/ray/python/ray/data/datasource/file_based_datasource.py", line 225, in _read_stream
(_execute_read_task pid=2023)     yield self._read_file(f, path, **reader_args)
(_execute_read_task pid=2023)   File "/Users/bveeramani/GitHub/ray/python/ray/data/datasource/image_folder_datasource.py", line 144, in _read_file
(_execute_read_task pid=2023)     if mode is not None:
(_execute_read_task pid=2023)   File "/Users/bveeramani/GitHub/ray/.venv/lib/python3.9/site-packages/PIL/Image.py", line 1035, in convert
(_execute_read_task pid=2023)     im = im.convert(mode, dither)
(_execute_read_task pid=2023) ValueError: conversion from L to X not supported
(_execute_read_task pid=2023) 2022-08-01 10:40:42,145   INFO worker.py:754 -- Task failed with retryable exception: TaskID(3d3e27c54ed1f5cfffffffffffffffffffffffff01000000).
(_execute_read_task pid=2023) Traceback (most recent call last):
(_execute_read_task pid=2023)   File "/Users/bveeramani/GitHub/ray/.venv/lib/python3.9/site-packages/PIL/Image.py", line 1030, in convert
(_execute_read_task pid=2023)     im = self.im.convert(mode, dither)
(_execute_read_task pid=2023) ValueError: conversion from 1 to X not supported
(_execute_read_task pid=2023) 
(_execute_read_task pid=2023) During handling of the above exception, another exception occurred:
(_execute_read_task pid=2023) 
(_execute_read_task pid=2023) Traceback (most recent call last):
(_execute_read_task pid=2023)   File "python/ray/_raylet.pyx", line 662, in ray._raylet.execute_task
(_execute_read_task pid=2023)     with ray._private.worker._changeproctitle(title, next_title):
(_execute_read_task pid=2023)   File "python/ray/_raylet.pyx", line 666, in ray._raylet.execute_task
(_execute_read_task pid=2023)     outputs = function_executor(*args, **kwargs)
(_execute_read_task pid=2023)   File "/Users/bveeramani/GitHub/ray/python/ray/data/_internal/lazy_block_list.py", line 541, in _execute_read_task
(_execute_read_task pid=2023)     block = task()
(_execute_read_task pid=2023)   File "/Users/bveeramani/GitHub/ray/python/ray/data/datasource/datasource.py", line 202, in __call__
(_execute_read_task pid=2023)     for block in result:
(_execute_read_task pid=2023)   File "/Users/bveeramani/GitHub/ray/python/ray/data/datasource/file_based_datasource.py", line 429, in read_files
(_execute_read_task pid=2023)     for data in read_stream(f, read_path, **reader_args):
(_execute_read_task pid=2023)   File "/Users/bveeramani/GitHub/ray/python/ray/data/datasource/file_based_datasource.py", line 225, in _read_stream
(_execute_read_task pid=2023)     yield self._read_file(f, path, **reader_args)
(_execute_read_task pid=2023)   File "/Users/bveeramani/GitHub/ray/python/ray/data/datasource/image_folder_datasource.py", line 144, in _read_file
(_execute_read_task pid=2023)     if mode is not None:
(_execute_read_task pid=2023)   File "/Users/bveeramani/GitHub/ray/.venv/lib/python3.9/site-packages/PIL/Image.py", line 1035, in convert
(_execute_read_task pid=2023)     im = im.convert(mode, dither)
(_execute_read_task pid=2023) ValueError: conversion from L to X not supported
(_execute_read_task pid=2023) 2022-08-01 10:40:43,144   INFO worker.py:754 -- Task failed with retryable exception: TaskID(3d3e27c54ed1f5cfffffffffffffffffffffffff01000000).
(_execute_read_task pid=2023) Traceback (most recent call last):
(_execute_read_task pid=2023)   File "/Users/bveeramani/GitHub/ray/.venv/lib/python3.9/site-packages/PIL/Image.py", line 1030, in convert
(_execute_read_task pid=2023)     im = self.im.convert(mode, dither)
(_execute_read_task pid=2023) ValueError: conversion from 1 to X not supported
(_execute_read_task pid=2023) 
(_execute_read_task pid=2023) During handling of the above exception, another exception occurred:
(_execute_read_task pid=2023) 
(_execute_read_task pid=2023) Traceback (most recent call last):
(_execute_read_task pid=2023)   File "python/ray/_raylet.pyx", line 662, in ray._raylet.execute_task
(_execute_read_task pid=2023)     with ray._private.worker._changeproctitle(title, next_title):
(_execute_read_task pid=2023)   File "python/ray/_raylet.pyx", line 666, in ray._raylet.execute_task
(_execute_read_task pid=2023)     outputs = function_executor(*args, **kwargs)
(_execute_read_task pid=2023)   File "/Users/bveeramani/GitHub/ray/python/ray/data/_internal/lazy_block_list.py", line 541, in _execute_read_task
(_execute_read_task pid=2023)     block = task()
(_execute_read_task pid=2023)   File "/Users/bveeramani/GitHub/ray/python/ray/data/datasource/datasource.py", line 202, in __call__
(_execute_read_task pid=2023)     for block in result:
(_execute_read_task pid=2023)   File "/Users/bveeramani/GitHub/ray/python/ray/data/datasource/file_based_datasource.py", line 429, in read_files
(_execute_read_task pid=2023)     for data in read_stream(f, read_path, **reader_args):
(_execute_read_task pid=2023)   File "/Users/bveeramani/GitHub/ray/python/ray/data/datasource/file_based_datasource.py", line 225, in _read_stream
(_execute_read_task pid=2023)     yield self._read_file(f, path, **reader_args)
(_execute_read_task pid=2023)   File "/Users/bveeramani/GitHub/ray/python/ray/data/datasource/image_folder_datasource.py", line 144, in _read_file
(_execute_read_task pid=2023)     if mode is not None:
(_execute_read_task pid=2023)   File "/Users/bveeramani/GitHub/ray/.venv/lib/python3.9/site-packages/PIL/Image.py", line 1035, in convert
(_execute_read_task pid=2023)     im = im.convert(mode, dither)
(_execute_read_task pid=2023) ValueError: conversion from L to X not supported
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/bveeramani/GitHub/ray/python/ray/data/read_api.py", line 312, in read_datasource
    block_list.ensure_metadata_for_first_block()
  File "/Users/bveeramani/GitHub/ray/python/ray/data/_internal/lazy_block_list.py", line 336, in ensure_metadata_for_first_block
    metadata = ray.get(metadata_ref)
  File "/Users/bveeramani/GitHub/ray/python/ray/_private/client_mode_hook.py", line 105, in wrapper
    return func(*args, **kwargs)
  File "/Users/bveeramani/GitHub/ray/python/ray/_private/worker.py", line 2245, in get
    raise value.as_instanceof_cause()
ray.exceptions.RayTaskError(ValueError): ray::_execute_read_task() (pid=2023, ip=127.0.0.1)
ValueError: conversion from 1 to X not supported

During handling of the above exception, another exception occurred:

ray::_execute_read_task() (pid=2023, ip=127.0.0.1)
  File "/Users/bveeramani/GitHub/ray/python/ray/data/_internal/lazy_block_list.py", line 541, in _execute_read_task
    block = task()
  File "/Users/bveeramani/GitHub/ray/python/ray/data/datasource/datasource.py", line 202, in __call__
    for block in result:
  File "/Users/bveeramani/GitHub/ray/python/ray/data/datasource/file_based_datasource.py", line 429, in read_files
    for data in read_stream(f, read_path, **reader_args):
  File "/Users/bveeramani/GitHub/ray/python/ray/data/datasource/file_based_datasource.py", line 225, in _read_stream
    yield self._read_file(f, path, **reader_args)
  File "/Users/bveeramani/GitHub/ray/python/ray/data/datasource/image_folder_datasource.py", line 144, in _read_file
    if mode is not None:
  File "/Users/bveeramani/GitHub/ray/.venv/lib/python3.9/site-packages/PIL/Image.py", line 1035, in convert
    im = im.convert(mode, dither)
ValueError: conversion from L to X not supported
>>> (_execute_read_task pid=2023) 2022-08-01 10:40:44,148       INFO worker.py:754 -- Task failed with retryable exception: TaskID(3d3e27c54ed1f5cfffffffffffffffffffffffff01000000).
(_execute_read_task pid=2023) Traceback (most recent call last):
(_execute_read_task pid=2023)   File "/Users/bveeramani/GitHub/ray/.venv/lib/python3.9/site-packages/PIL/Image.py", line 1030, in convert
(_execute_read_task pid=2023)     im = self.im.convert(mode, dither)
(_execute_read_task pid=2023) ValueError: conversion from 1 to X not supported
(_execute_read_task pid=2023) 
(_execute_read_task pid=2023) During handling of the above exception, another exception occurred:
(_execute_read_task pid=2023) 
(_execute_read_task pid=2023) Traceback (most recent call last):
(_execute_read_task pid=2023)   File "python/ray/_raylet.pyx", line 662, in ray._raylet.execute_task
(_execute_read_task pid=2023)     with ray._private.worker._changeproctitle(title, next_title):
(_execute_read_task pid=2023)   File "python/ray/_raylet.pyx", line 666, in ray._raylet.execute_task
(_execute_read_task pid=2023)     outputs = function_executor(*args, **kwargs)
(_execute_read_task pid=2023)   File "/Users/bveeramani/GitHub/ray/python/ray/data/_internal/lazy_block_list.py", line 541, in _execute_read_task
(_execute_read_task pid=2023)     block = task()
(_execute_read_task pid=2023)   File "/Users/bveeramani/GitHub/ray/python/ray/data/datasource/datasource.py", line 202, in __call__
(_execute_read_task pid=2023)     for block in result:
(_execute_read_task pid=2023)   File "/Users/bveeramani/GitHub/ray/python/ray/data/datasource/file_based_datasource.py", line 429, in read_files
(_execute_read_task pid=2023)     for data in read_stream(f, read_path, **reader_args):
(_execute_read_task pid=2023)   File "/Users/bveeramani/GitHub/ray/python/ray/data/datasource/file_based_datasource.py", line 225, in _read_stream
(_execute_read_task pid=2023)     yield self._read_file(f, path, **reader_args)
(_execute_read_task pid=2023)   File "/Users/bveeramani/GitHub/ray/python/ray/data/datasource/image_folder_datasource.py", line 144, in _read_file
(_execute_read_task pid=2023)     if mode is not None:
(_execute_read_task pid=2023)   File "/Users/bveeramani/GitHub/ray/.venv/lib/python3.9/site-packages/PIL/Image.py", line 1035, in convert
(_execute_read_task pid=2023)     im = im.convert(mode, dither)
(_execute_read_task pid=2023) ValueError: conversion from L to X not supported
```

Reply:
I think it's fine now, thanks @bveeramani for checking.